### PR TITLE
programs.neovim: Add initExtra option (2)

### PR DESCRIPTION
**Pull request by viniciusmuller**
### Description
The last PR was truly messy (#2244) so I've closed it and I'm opening a new one now.

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

ping @berbiche

### CODE REVIEW

**Comment by teto** [comment]
With neovim we have people writing their config in lua/fennel/teal; I would like to supporting these approaches in neovim. This means we may have to modify the option aftwards. @berbiche I'll let you handle the reviews.

**Comment by kclejeune** [comment]
> With neovim we have people writing their config in lua/fennel/teal; I would like to supporting these approaches in neovim. This means we may have to modify the option aftwards. @berbiche I'll let you handle the reviews.

Maybe I'm missing something but I always assumed that the easiest solution to this in the future would be to add an option specifying the output filetype. We can reuse the existing `initExtra` option but put it into `init.lua` or some other file format instead.

**Comment by teto** [comment]
I would be happy for someone to take on this but we should agree on the design first. 
My current take is: 
- to add a freeform "language" option to the plugin submodule (that would default to vim)
- add a function that collects config with same language and does what it wants with it. By default it would write to init.LANGUAGE

As for the initExtra, someone mentioned putting its initExtra as a fake plugin. Looks like an interesting workaround.

I am scared to add something to complex for users and/or to maintain. At the same time as someone with a complex hybrid lua/viml config (and looking forward to add some fennel), I would still like to make it possible for advanced users.

File: modules/programs/neovim.nix [hunk 1]
```diff
@@ -54,6 +54,11 @@ let
           cfg.plugins);
     };
     beforePlugins = "";
+  } // optionalAttrs (cfg.initExtra != "") {
+    beforePlugins = ''
+      " Generated by Home Manager
+        ${cfg.initExtra}
```
Comment (expipiplus1) [inline modules/programs/neovim.nix:60]: Indented here, but not in the test case?

**Comment by thiagokokada** [comment]
I was bitten by this issue after `release-21.11`. I remap my leader from `\` to `Space`, and since it seems `programs.neovim.extraConfig` is called later than the plugins my config stopped working (it used to work before).

Could we have this PR merged and backport to `release-21.11`? It maybe not perfect but at least it solves the issue.

**Comment by GlancingMind** [comment]
> I was bitten by this issue after `release-21.11`. I remap my leader from `\` to `Space`, and since it seems `programs.neovim.extraConfig` is called later than the plugins my config stopped working (it used to work before).

For now I work around the same Issue by prepanding the first plugins configuration with my nvimrc. E.g.
```
programs.neovim = {
  ...
  plugins = [
    {
      plugin = pkgs.vimPlugins.gruvbox;
      config = builtins.readFile ./config/nvimrc + builtins.readFile ./config/gruvbox.vim;
    }
  ];
};
```

**Comment by berbiche** [comment]
@teto following the merge of https://github.com/nix-community/home-manager/pull/2637, what should be changed here to have this merged?

**Comment by teto** [comment]
I am annoyedby the `set nocompatible` that creeps in the generated init.vim .
Also now that plugins can specify the language of their config, it would be nice to be able to precise this in option, @selene57 might have a go at this https://github.com/nix-community/home-manager/pull/2716

**Comment by berbiche** [comment]
> @teto I am annoyedby the `set nocompatible` that creeps in the generated init.vim . Also now that plugins can specify the language of their config, it would be nice to be able to precise this in option, @selene57 might have a go at this #2716

If we drop the the `set nocompatible` line and keep the Home Manager comment, do you think we could merge this?

**Review by teto — APPROVED** [review]

**Comment by expipiplus1** [comment]
Is this still on the table?

**Comment by Wisehh** [comment]
> > I was bitten by this issue after `release-21.11`. I remap my leader from `\` to `Space`, and since it seems `programs.neovim.extraConfig` is called later than the plugins my config stopped working (it used to work before).
> 
> For now I work around the same Issue by prepanding the first plugins configuration with my nvimrc. E.g.
> 
> ```
> programs.neovim = {
>   ...
>   plugins = [
>     {
>       plugin = pkgs.vimPlugins.gruvbox;
>       config = builtins.readFile ./config/nvimrc + builtins.readFile ./config/gruvbox.vim;
>     }
>   ];
> };
> ```

Thank you for sharing your answer, it saved me a lot of time.
In case it helps anyone else, if you have your entire configuration defined in an init.lua, you can separate it into 2 groups, the "core" configurations (that you need to run at the beginning) and the plugins config. That done, you will end up with something like:

```
programs.neovim = {
      ...
      extraConfig = ":luafile ~/.config/nvim/append_config.lua";
      plugins = with pkgs.vimPlugins; [
        { plugin = gruvbox-nvim;
          config = ("lua << EOF\n" + builtins.readFile ./nvim/preppend_config.lua + "EOF");
        }
       ...
       nvim-cmp
       etc
        ...
     ];
};
```
Where preppend_config will load the core, and append_config will take care of the plugins at the end.

However, this feels quite convoluted. would it be possible for you to please reconsider the initExtra option? I really consider it an option that carries enough value to create it!

Anyway, thank you and best regards!

**Comment by expipiplus1** [comment]
FWIW, I use this almost trivial patch. If someone wants to neaten it up (change the name to something more descriptive than `initExtra`, allow it to accept lua, add tests) then please feel free to take it. 

```diff
diff --git a/modules/programs/neovim.nix b/modules/programs/neovim.nix
index df4b487d..28abf1e0 100644
--- a/modules/programs/neovim.nix
+++ b/modules/programs/neovim.nix
@@ -98,6 +98,14 @@ in {
     programs.neovim = {
       enable = mkEnableOption "Neovim";

+      initExtra = mkOption {
+        type = types.path;
+        default = null;
+        description = ''
+          Configuration to insert at the top of <filename>$XDG_CONFIG_HOME/nvim/init.vim</filename>
+        '';
+      };
+
       viAlias = mkOption {
         type = types.bool;
         default = false;
@@ -401,6 +409,8 @@ in {
         (map (x: x.runtime) pluginsNormalized) ++ [{
           "nvim/init.lua" = let
             luaRcContent =
+              lib.optionalString (cfg.initExtra != null)
+              "vim.cmd [[source ${cfg.initExtra}]]\n" +
               lib.optionalString (neovimConfig.neovimRcContent != "")
               "vim.cmd [[source ${
                 pkgs.writeText "nvim-init-home-manager.vim"
```

**Comment by ncfavier** [comment]
Why not `extraConfig = mkBefore ...`?

**Comment by expipiplus1** [comment]
@ncfavier because `extraConfig` itself is always placed after the plugin configs, so even if something is at the top of `extraConfig` it's too late.

File: tests/modules/programs/neovim/plugin-config.vim [hunk 2]
```diff
@@ -1,3 +1,7 @@
+" generated by Home Manager
+set nocompatible
```
Comment (teto) [inline tests/modules/programs/neovim/plugin-config.vim:2]: neovim sets nocompatible by default so it shoulkd not appear

**Comment by teto** [comment]
ideally nixpkgs' code should be rewritten with the nixos module so that we dont have to create extra options and can just do mkBefore/mkAfter.

I am counting on https://github.com/NixOS/nixpkgs/pull/237425 to make overrides easier too.

**Comment by teto** [comment]
dont update this PR yet, there is another approach I would like to take now that most of it is merged in nixpkgs.

**Comment by thiagokokada** [comment]
@teto Any updates here?

**Comment by teto** [comment]
this will become obsolete once https://github.com/nix-community/home-manager/pull/5964 gets merged (after the 24.11 release)

**Pull request closed without merging**